crates/ruff_linter/src/linter.rs

```diff
@@ -214,6 +215,39 @@ pub fn check_path(
         ));
     }
 
+    // Raise violations for internal test rules
+    #[cfg(feature = "test-rules")]
+    {
+        for test_rule in TEST_RULES {
+            if settings.rules.enabled(*test_rule) {
+                let diagnostic = match test_rule {
+                    Rule::StableTestRule => {
+                        test_rules::StableTestRule::diagnostic(&locator, &indexer)
+                    }
+                    Rule::StableTestRuleSafeFix => {
+                        test_rules::StableTestRuleSafeFix::diagnostic(&locator, &indexer)
+                    }
+                    Rule::StableTestRuleUnsafeFix => {
+                        test_rules::StableTestRuleUnsafeFix::diagnostic(&locator, &indexer)
+                    }
+                    Rule::StableTestRuleDisplayOnlyFix => {
+                        test_rules::StableTestRuleDisplayOnlyFix::diagnostic(&locator, &indexer)
+                    }
+                    Rule::NurseryTestRule => {
+                        test_rules::NurseryTestRule::diagnostic(&locator, &indexer)
+                    }
+                    Rule::PreviewTestRule => {
+                        test_rules::PreviewTestRule::diagnostic(&locator, &indexer)
+                    }
+                    _ => unreachable!("All test rules must have an implementation"),
+                };
+                if let Some(diagnostic) = diagnostic {
+                    diagnostics.push(diagnostic);
+                }
+            }
+        }
```

Comment:
Why is this so awkward Rust... ?

Reply:
I'd probably do

```rust
if !settings.rules.enabled(*test_rule) {
    continue;
}
```

but it otherwise doesn't look too bad?

Reply:
I don't get why I can't call `::diagnostic` once since I know all of the objects implement `TestRule`.

I have to do some weird dynamic boxing to get to that point?

Reply:
I don't follow. Can you show me the code you _want_ to write? It looks like you're calling a bunch of different impls of `diagnostic` based on which `Rule` you have? Each line is doing the same function call, but it's on a different type.

Reply:
`diagnostic` is provided by the `TestRule` trait which all of these rules implement. For a while I had...

```rust
        for test_rule in TEST_RULES {
            if settings.rules.enabled(*test_rule) {
                if let Some(diagnostic) =
                    implementation_for_test_rule(*test_rule).diagnostic(&locator, &indexer)
                {
                    diagnostics.push(diagnostic);
                }
            }
        }
```

but that required the slightly awkward

```rust
pub(crate) fn implementation_for_test_rule(rule: Rule) -> Box<dyn TestRule>{
    match rule {
        Rule::StableTestRule => Box::new(StableTestRule),
        Rule::StableTestRuleSafeFix => Box::new(StableTestRuleSafeFix),
        Rule::StableTestRuleUnsafeFix => Box::new(StableTestRuleUnsafeFix),
        Rule::StableTestRuleDisplayOnlyFix => Box::new(StableTestRuleDisplayOnlyFix),
        Rule::NurseryTestRule => Box::new(NurseryTestRule),
        Rule::PreviewTestRule => Box::new(PreviewTestRule),
        _ => unreachable!("All test rules must have an implementation"),
    }
}

```

Reply:
I was like.. why am I importing `TEST_RULES` and `implementation_for_test_rule` from `test_rules` I should just do something like `iter_test_rules() -> &[Rule, Box<dyn TestRule>]` and that's when I ran into horrible cyclical compiler errors

Reply:
I think any way you slice things here, you need a match statement over a `Rule`. I don't see a way around that because something has to tell the compiler which impl of `TestRule` to use, and _that_ is determined by the `Rule`. Even if you wrote `iter_test_rules() -> &[Rule, Box<dyn TestRule>]`, you'd still need a match statement in there.

I think where you ended up is fine IMO.

Reply:
Okay thanks!